api/dotnet/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+
+        let val_str = format!("{:?}", val);
+        ret.type_name = name.into();
+        ret.type_type = val.value_type() as i32;
+        ret.type_value = val_str.clone();
+
+        printdebug!("{}", val_str);
+    });
+
+    ret
+}
+
+#[net]
+pub fn get_callbacks() -> Vec<String> {
+    printdebug!("get_callbacks()");
+
+    let mut ret: Vec<String> = Vec::new();
+
+    CURRENT_INSTANCE.with(|current| {
+        let strong_ref = current.borrow_mut().take().unwrap();
+        current.replace(Some(strong_ref.clone_strong()));
+
+        let binding = strong_ref.definition();
+        let calls = binding.callbacks();
+        ret = calls.collect();
+    });
+
+    ret
+}
+
+#[net]
+pub fn set_callback(name: &str, callback: Delegate0<bool>) {
+    printdebug!("set_callback()");
+
+    CURRENT_INSTANCE.with(|current| {
+        let strong_ref = current.borrow_mut().take().unwrap();
+        current.replace(Some(strong_ref.clone_strong()));
+
+        strong_ref
+            .set_callback(name, move |_| {
+                callback.call();
+                Value::Void
+            })
+            .unwrap();
+    });
+}
+
+#[net]
+pub fn new_timer(mode: i32, interval: u64, callback: Delegate0<bool>) -> DotNetTimer {
+    let ret = DotNetTimer { timer_id: -1, interval };
+
+    TIMER_POOL.with(|pool| {
+        let mut mut_pool = pool.borrow_mut().take().unwrap();
+        let timer = Timer::default();
+
+        let time_mode = if mode == 1 { TimerMode::Repeated } else { TimerMode::SingleShot };
+
+        let int_duration = Duration::from_millis(interval);
+
+        timer.start(time_mode, int_duration, move || {
+            callback.call();
+        });
+
+        mut_pool.push(timer);
+        pool.replace(Some(mut_pool));
+    });
+
+    ret
+}
+
+#[net]
+pub fn stop_timer(timer: DotNetTimer) {
+    TIMER_POOL.with(|pool| {
+        let pool = pool.borrow_mut().take().unwrap();
+        let mut index = 0;
+
+        // for each timer check if the id match
+        for tmr in pool {
+            if index == timer.timer_id {
+                tmr.stop();
+                break;
+            }
+
+            index += 1;
+        }
+    });
+}
+
+#[net]
+pub fn restart_timer(timer: DotNetTimer) {
+    TIMER_POOL.with(|pool| {
+        let pool = pool.borrow_mut().take().unwrap();
+        let mut index = 0;
+
+        // for each timer check if the id match
+        for tmr in pool {
+            if index == timer.timer_id {
+                tmr.restart();
+                break;
+            }
+
+            index += 1;
+        }
+    });
+}
+
+#[net]
+pub fn run_on_ui_thread(callback: Delegate0<bool>) {
+    printdebug!("run_on_ui_thread()");
+
+    // reject modernity, back to the monke
+    let weak_ref = unsafe { MAIN_WEAK_INSTANCE.take().unwrap() };
+    unsafe {
+        MAIN_WEAK_INSTANCE = Some(weak_ref.clone());
+    };
+
+    weak_ref
+        .upgrade_in_event_loop(move |_| {
+            callback.call();
+        })
```

Comment:
You can use `slint_interpreter::invoke_from_event_loop` , no need to have a weak_ref.  That way you can get rid of this `MAIN_WEAK_INSTANCE` entirely